oldScripts/pl_to_spreadsheet.py

```diff
@@ -0,0 +1,102 @@
+# Author: Naveen Nathan
+import time, requests, datetime
+
+import os.path
+import json
+import pandas as pd
+from dotenv import load_dotenv
+"""
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+"""
+
+# CS10 Su24 course id
+COURSE_INSTANCE_ID = 155812
+SERVER = "https://us.prairielearn.com/pl/api/v1"
+
+# This scope allows for write access.
+SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]
+SPREADSHEET_ID = "16e6hWK4wWiqetuyJrDvBy4O9Wwp83NR1JptZU1dIIxI"
+GRADEBOOK_SPREADSHEET_ID = ""
+
+load_dotenv()
+PL_API_TOKEN = os.getenv("PL_API_TOKEN")
+
+"""
+Allows the user authenticate their google account, allowing the script to modify spreadsheets in their name.
+Borrowed from here: https://developers.google.com/sheets/api/quickstart/python
+"""
+
+
+def allow_user_to_authenticate_google_account():
+    creds = None
+    # The file token.json stores the user's access and refresh tokens, and is
+    # created automatically when the authorization flow completes for the first
+    # time.
+    if os.path.exists("token.json"):
+        creds = Credentials.from_authorized_user_file("token.json", SCOPES)
+    # If there are no (valid) credentials available, let the user log in.
+    if not creds or not creds.valid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = InstalledAppFlow.from_client_secrets_file(
+                "credentials.json", SCOPES
+            )
+            creds = flow.run_local_server(port=0)
+            print("Authentication succesful")
+    # Save the credentials for the next run
+    with open("token.json", "w") as token:
+        token.write(creds.to_json())
+    return creds
+
+
+"""
+This method is adapted from Prairielearn's public repository.
+https://github.com/PrairieLearn/PrairieLearn/blob/63c90a6523a3061743b8653a4cfafc62e0e0dbff/tools/api_download.py#L68
+"""
+
+
+def get_json(endpoint, token):
+    url = SERVER + endpoint
+    headers = {'Private-Token': token}
+    start_time = time.time()
+    retry_502_max = 30
```

Comment:
I recommend using an exponential backoff mechanism. This will abstract the retry logic away and there is no need for a while True loop. You can look into other libraries, but here is one https://backoff-utils.readthedocs.io/en/latest/

Reply:
Thanks for suggesting this, @Mehul-Gandhi! Glad to learn; will use this.